### PR TITLE
throttle

### PR DESCRIPTION
**Pull request by Uzlopak**
#### Checklist

- [ ] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Review by mcollina — CHANGES_REQUESTED** [review]
You should use pipeline instead, .pipe does not add error handling.

Note that using an intermediate stream adds significant overhead.

**Comment by Uzlopak** [comment]
> You should use pipeline instead, .pipe does not add error handling.

Ok, I will use pipeline

> Note that using an intermediate stream adds significant overhead.

Should I put that into the readme? Or is there a different approach, that I am not aware of?

I actually wanted to write about the feature for the blog of @Eomm 

https://github.com/Uzlopak/blog/blob/main/case.md

**Comment by Uzlopak** [comment]
@mcollina 

Do you think it makes sense to create a new package called @fastify/throttle?

File: index.js [hunk 1]
```diff
@@ -158,6 +170,41 @@ async function fastifyRateLimit (fastify, settings) {
...
+  const bps = throttleOpts.bps
+  return function (request, reply, payload, done) {
+    if (payload && payload.pipe) {
+      const throttleStream = new ThrottleStream({ bps })
+      payload.pipe(throttleStream)
+      done(null, throttleStream)
+      return
+    } else if (Buffer.isBuffer(payload)) {
+      const throttleStream = new ThrottleStream({ bps })
+      Readable.from(payload).pipe(throttleStream)
```
Comment (mcollina) [inline index.js:195]: It does not have much sense to throttle in this case. Throlling is needed to limit the resource consumed by the caller. In this specific case, we are holding quite a bit of data in memory, literally increasing the resource consumed to process the request.

**Review by mcollina — COMMENTED** [review]
I think a new module would be better.

---

Something that this does not handle is throttling across multiple request/multiple server for a given user.

**Comment by Uzlopak** [comment]
I am actually working on the distributed part. 

see
https://github.com/Uzlopak/blog/blob/main/case.md#throttlegroups

also yes, i initially wanted to just handle streams. Maybe make it optional to handle also buffer and strings if an option is set (opt-in)

**Comment by Uzlopak** [comment]
Created a new repo fastify/throttle and will do the changes there.

**Pull request closed without merging**